archive/miu_language/decision_nec.lean

```diff
@@ -0,0 +1,310 @@
+derivable string must be 1 or 2 moduloe 3. To do this, it suffices to show that if the `miustr`
+`en` is derived from `st`, then `icount en` moudulo 3 is either equal to or is twice `icount st`
+modulo 3.
+-/
+
+/--
+`icount` is the number of `I`s in an `miustr`
+-/
+def icount : miustr → ℕ
+| [] := 0
+| (I::cs) := 1 + icount cs
+| (_::cs) := icount cs
+
+/--
+`icount` is additive with respect to `append`
+-/
+lemma icountappend (a b : miustr) :
+  icount (a ++ b) = icount a + icount b :=
+begin
+  induction a with x hx hxs, -- treat a as x :: xs in the inductive step
+    simp [icount], -- trivial base case
+    cases x; -- the same proof applies whether x = 'M', 'I', or 'U'
+      simp [icount, hxs, add_assoc],
+end
+
+open nat
+
+/--
+Given `st en : miustr`, the relation `nice_imod3 st en` holds if `st` and `en` either
+have equal `icount`, modulo 3, or `icount en` is twice `icount st`, modulo 3.`
+ -/
+def nice_imod3 (st en : miustr) : Prop :=
+  let a := (icount st) in
+  let b := (icount en) in
+  b ≡ a [MOD 3] ∨ b ≡ 2*a [MOD 3]
+
+example : nice_imod3 "II" "MIUI" :=
+begin
+  left, refl, -- icount "MIUI" ≡ icount "II" [MOD 3]
+end
+
+example : nice_imod3 "IUIM" "MI" :=
+begin
+  right, refl, --icount "MI" ≡ 2*(icount "IUIUM") [MOD 3]
+end
+
+
+/-  We show the icount, mod 3, stays the same or is multiplied by 2 under the rules of inference -/
+
+open nat
+
+/- Now we show that the icount of a derivable string is 1 or 2 modulo 3-/
+
+-- We start with a general result about natural numbers.
+
+lemma inheritmod3 {a b : ℕ} (h1 : a % 3 = 1 ∨ a % 3 = 2)
+  (h2 : b % 3 = a % 3 ∨  b % 3 = (2 * a % 3)) :
+    b % 3 = 1 ∨ b % 3 = 2 :=
+begin
+  cases h2, {
+    rw h2,
+    exact h1,
+  }, {
+    cases h1, {
+      right,
+      simp [h2,mul_mod,h1],
+      refl,
+    }, {
+      left,
+      simp [h2,mul_mod,h1],
+      refl,
+    }
+  }
+end
+
+
+/--
+`inctder` shows any derivable string must have an `icount` that is 1 or 2 modulo 3.
+-/
+theorem icntder (en : miustr): derivable en →
+  (icount en) % 3 = 1 ∨ (icount en) % 3 = 2:=
+begin
+  intro h,
+  induction h,
+    left,
+    apply mod_def,
+    any_goals {apply inheritmod3 h_ih},
+      left, simp only [icountappend], refl,
+      right, simp only [icount,icountappend],ring,
+      left, simp [icountappend,icount,refl], ring,
+      left, simp [icountappend,icount,refl],
+end
+
+/--
+Using the above theorem, we solve the MU puzzle, showing that `"MU"` is not derivable.
+-/
+theorem notmu : ¬(derivable "MU") :=
+begin
+  intro h,
+  cases (icntder _ h);
+    contradiction,
+end
+
+
+/-
+### Condition on `M`
+
+That solves the MU puzzle, but we'll proceed by demonstrating the other necessary condition for a
+string to be derivable, namely that the string must start with an M and contain no M in its tail.
+-/
+
+
+/--
+`goodm xs` holds if `xs : miustr` begins with `M` and contains no `M` in its tail.
+-/
+def goodm (xs : miustr) : Prop :=
+  ∃ ys : miustr, xs = (M::ys) ∧ ¬(M ∈ ys)
```

Comment:
Is it necessary to use an existential here? Can you just use `list.head` and `list.tail`?

Reply:
I did that at first but ran into the issue that the empty list has a head.

Reply:
I guess it is a bit gross to rely on the `default` element not being `M`. :-) Okay!

Reply:
Hmm, I've taken on board this suggestion (with a bit of work), and I get `decidable_pred decstr` via the magic `@derive'. I still need to rewrite the `decision_suf.lean` file to work with the new definition.

Once done, I should be able to prove `decidable_pred derivable`.

Reply:
OK. I think I'm done now. Have made the change and thereby proved `decidable_pred derivable`.